### ISSUE TITLE
my phone will

### ISSUE THREAD

**Issue by arbaleast** [wa2c/cifs-documents-provider#35]

**Comment by arbaleast** [comment]
os: android10
storage: smb(jcifs-ng)
the app: poweramp
and what info should I provide?

**Comment by arbaleast** [comment]
thanks in advance

**Comment by wa2c** [comment]
I don't know the cause.
Please tell me the model name for reference.

**Comment by arbaleast** [comment]
![Screenshot_20230516-202117_(1)](https://github.com/wa2c/cifs-documents-provider/assets/57560360/7a2fd3c0-f430-4a9f-8b39-f85a755e4d5f)
sorry, late to reply

> I don't know the cause. Please tell me the model name for reference.

**Comment by user-gac** [comment]
同様の現象が発生しています

## Issue
Androidが再起動する(=AndroidRuntimeが再起動する)

## Environment
	Pixel 7 Pro
	 Android 13
	  CIFS Documents Provider 1.70
	  PowerAmp build-967
	Windows 8

## Reproducing
1. Windows の共有フォルダに1031ファイル(39フォルダ)ほどの音楽データを格納する
2. CIFS Documents Provider にて上記フォルダに正しく接続できる情報を一つ設定する(安全なデータ転送付き)
   \# このストレージ接続が正しく動作しているか確認するためにMiXplorerを使用した
3. PowerAmp の ライブラリ として 上記 CIFS Documents Providerで設定したストレージを追加する
   \# 自動でフル再スキャンされる。または手動でフル再スキャンを行う
4. スキャン最中にスキャンが進行しなくなり(固まり)、その後何十秒後かにAndroidが再起動する
   \# いつ止まるかは不規則(数十～数百ファイルあたり)
5. この際、`logcat -b crash,system` を流していると、再起動の瞬間あたりで以下のログが表示される
```
08-05 02:34:11.115  7798  7821 W Watchdog: *** WATCHDOG KILLING SYSTEM PROCESS: Blocked in monitor com.android.server.StorageManagerService on foreground thread (android.fg)
08-05 02:34:11.116  7798  7821 W Watchdog: android.fg annotated stack trace:
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.BinderProxy.transactNative(Native Method)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.BinderProxy.transact(BinderProxy.java:584)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.IVold$Stub$Proxy.monitor(IVold.java:1470)
08-05 02:34:11.116  7798  7821 W Watchdog:     at com.android.server.StorageManagerService.monitor(StorageManagerService.java:4771)
08-05 02:34:11.116  7798  7821 W Watchdog:     at com.android.server.Watchdog$HandlerChecker.run(Watchdog.java:330)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.Handler.handleCallback(Handler.java:942)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.Handler.dispatchMessage(Handler.java:99)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.Looper.loopOnce(Looper.java:201)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.Looper.loop(Looper.java:288)
08-05 02:34:11.116  7798  7821 W Watchdog:     at android.os.HandlerThread.run(HandlerThread.java:67)
08-05 02:34:11.116  7798  7821 W Watchdog:     at com.android.server.ServiceThread.run(ServiceThread.java:44)
08-05 02:34:11.116  7798  7821 W Watchdog: *** GOODBYE!
08-05 02:34:11.471   415  7455 E vold    : Failed to open AppFuse file for uid: 10589 mountId: 58 fileId: 2 flags: 0: Software caused connection abort
08-05 02:34:11.587  8669  8669 E AndroidRuntime: FATAL EXCEPTION: main
08-05 02:34:11.587  8669  8669 E AndroidRuntime: Process: com.google.android.gms.persistent, PID: 8669
08-05 02:34:11.587  8669  8669 E AndroidRuntime: DeadSystemException: The system died; earlier logs will point to the root cause
```

- jcifs-ng、SMBJ どちらでも発生する
- CIFS Documents Provider 1.60 では発生したことが**ない**
- PowerAmp のフル再スキャンが再現に手っ取り早いだけで、他アプリを介したストレージアクセスでもAndroid再起動することがある

## Result expected
- 止まらず正しくフルスキャン(1031ファイル)を終える
- Androidが再起動しない

**Comment by wa2c** [comment]
I was able to reproduce this event.
We are investigating the cause.

**Comment by user-gac** [comment]
that sounds really great
hope itll be fixed soon! :)

**Comment by user-gac** [comment]
yay thats neat!
but wonder why it didnt occur in v1.60?
just curious.
cant wait next update

**Comment by RebelliousX** [comment]
It is a lot faster to scan lots of files,. For example, when using PPSSPP emulator, to scan just 42 games it used to take 4-5 minutes. Now It is around 1:20.. still slow, but faster. 

I still encounter random reboots when using PPSSPP (it doesn't happen after the initial scan of the games, but if I run PPSSPP again without rebooting the system, Android will crash and reboot during the second time PPSSPP will scan the games directory, a quick reboot). This happens on my Z Fold 3, Android emulator on Win11 and Xperia phone. Strangely, it is the only emulator with this issue.

Also, it seems SMBj, although It is greatly faster than jCIFS when it comes to scanning or reading files, it is not as reliable as jCIFS, I get read errors (corrupted files) when opening files like disc images.. I don't get this when using jCIFS, I would love to use SMBJ since it is faster for me.

Are we not supposed to have multiple connections active at the same time each is using different library SMBJ and JCIFS at the for different folders in the share path? Also, I get issues like mentioned above if I changed the connection from one library to the other after it was used.

After v1.8.0, the Files app sometimes says "Can't load content at the moment".. I have to pull down to refresh few times for the files to show up. It doesn't happen all the time. This seems to be related to the parallel file limitation set to 10, which I believe is the right step to fix this issue, but I also noticed there is a Time Out for each file opened which is set to 10'000? 

Can we have options added to the settings of the app to modify these values?

Maybe I should open another issue for my rants.

Thanks @wa2c again.

**Comment by user-gac** [comment]
woo word!
tried v1.8.0 on playstore
PowerAmp full scan completed without any problem 5 times in a row
no more ART rebooting
thanks for the fix :D

**Comment by dsgls** [comment]
I'm still seeing this issue on v2.4.0 with Android 16 QPR2. Triggered from USB Audio Player scanning but also with my own music player app (which buffers pretty aggressively).
```
01-08 16:28:22.877  1382 31144 I DropBoxManagerService: add tag=system_server_watchdog isTagEnabled=true flags=0x6
01-08 16:28:22.891  1382  1440 W Watchdog: *** WATCHDOG KILLING SYSTEM PROCESS: Blocked in monitor com.android.serve
r.StorageManagerService on monitor thread (watchdog.monitor) for 63s
01-08 16:28:22.893  1382  1440 W Watchdog: watchdog.monitor annotated stack trace:
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.BinderProxy.transactNative(Native Method)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.BinderProxy.transact(BinderProxy.java:626)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.IVold$Stub$Proxy.monitor(IVold.java:1624)
01-08 16:28:22.893  1382  1440 W Watchdog:     at com.android.server.StorageManagerService.monitor(StorageManagerSer
vice.java:4822)
01-08 16:28:22.893  1382  1440 W Watchdog:     at com.android.server.Watchdog$HandlerChecker.run(Watchdog.java:383)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.Handler.handleCallback(Handler.java:1070)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.Handler.dispatchMessage(Handler.java:125)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.Looper.dispatchMessage(Looper.java:333)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.Looper.loopOnce(Looper.java:263)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.Looper.loop(Looper.java:367)
01-08 16:28:22.893  1382  1440 W Watchdog:     at android.os.HandlerThread.run(HandlerThread.java:139)
01-08 16:28:22.893  1382  1440 W Watchdog:     at com.android.server.ServiceThread.run(ServiceThread.java:46)
01-08 16:28:22.893  1382  1440 W Watchdog: *** GOODBYE!
01-08 16:28:23.095   489   489 E vold    : Failed to open AppFuse file for uid: 10174 mountId: 740 fileId: 37 flags:
 0: Software caused connection abort
01-08 16:28:23.096 32207  8758 D ActivityThread: Too many transaction errors, throttling freezer binder callback.
01-08 16:28:23.123  2348  2348 D ActivityThread: Too many transaction errors, throttling freezer binder callback.
01-08 16:28:23.129  2348  2348 E AndroidRuntime: FATAL EXCEPTION: main
01-08 16:28:23.129  2348  2348 E AndroidRuntime: Process: com.android.phone, PID: 2348
01-08 16:28:23.129  2348  2348 E AndroidRuntime: DeadSystemException: The system died; earlier logs will point to the root cause
```

So it looks like any intensive file access through the CIFS provider causes vold to block and eventually the watchdog kills it. Since it's a system process this causes a reboot. Of course, the root of the problem lies in android itself - an app should not be able to lock up a system process.

**Comment by dsgls** [comment]
I opened a ticket on the android bug tracker. Don't expect much, but worth a shot.
https://issuetracker.google.com/issues/474311252

**Comment by RebelliousX** [comment]
@dsgls I can't reproduce the problem, I used PowerAmp to scan over 22+GB of OCRemix library (professional game music remixes, over 4000+ mp3 files), within 2 minutes. And since USB Audio Player Pro is a paid app, I downloaded the trial from the official website, and it scanned the library in the same amount of time or even less.

This is just a guess; the system process being killed (that manages storage) is due to the fact CIFS Document Provider app is killed in the background, when Android sees the app requesting file to be opened is not there anymore it kills any opened processes and try to free allocated memory and worst case, memory leaks causes Android to reboot (which could indicate an Android bug indeed). 

To prevent that, enable both options in the settings of the app to "act as local storage" and to "display notifications" which will use Android WorkManager to create foreground service to prevent the app getting terminated when resources are exhausted.  The foreground service is only active when a file is opened and stops when there are none. So, no battery drains whatsoever.

And note that I am using a slightly modified build of CIFS Document Provider for speed, you can try here if you want (it is based on the PR #100 which may or may not be merged. If you want to try it, follow the instructions in the release page for best possible performance 😊 

https://github.com/RebelliousX/cifs-documents-provider/releases

I am too lazy to try the original version, which is a lot slower when scanning myriads of files, but it should work ( I believe ) provided you enabled both mentioned settings and allow notification permission for the app from Android settings ( that is, you did not deny it when installed the app first time ).

Admittedly, I used to get phone reboots when scanning for game disc files in emulators (like PPSSPP and others) long time ago, but since the fixes for file limitation ( I kept it to default 30 ) and persistent foreground service to prevent app kill, I never encountered this issue anymore.

**Comment by dsgls** [comment]
@RebelliousX thanks for looking into it :)

The issue only happens intermittently. I can run long scans for hours without issue (my music player builds waveforms during indexing, so needs to read and decode every file), then out of the blue I'll see the phone reboot. So it's not easy to reproduce.

Your suggestion that the CIFS document provider is being killed is interesting. Do you know if android would log anything when that happens? I haven't seen anything like that, but the system log is super verbose so I might have missed it. But also my phone usage is very light so I can't imagine what would cause the system to run low on resources. The only things running apart from the music player and the CIFS document provider are background stuff from signal, DAVx5 and thunderbird. The phone has 12GiB RAM so unless there is a serious memory leak somewhere it shouldn't be an issue.

I'll try the local storage and notification settings and see if it happens again. Thanks again.

**Comment by RebelliousX** [comment]
@dsgls It is not about how much RAM you have, it all depends on Android operating system what it considers a running app should be killed or not. If you look in the Developer Options, there is an option "Background Process Limit" which is set to "Standard" by default, I can be wrong, but I read the standard is equivalent to 4 processes same to the max value. Android is not like Windows with its time sharing for CPU scheduling.

That doesn't mean Android will kill apps left and right even if running more than 4 processes, but I think if it notices an app is taking the lion share of the CPU for some time then it might deem other apps as inactive and suspends them. If CIFS Document Provider which is running in the background all the time is suspended while a file is opened, then you notice monkey business going on.

Long time ago, before the fix for the foreground service was introduced, I was running PPSSPP emulator on a Z Fold 3 phone that has 12GB Ram back then, even though PPSSPP is very lightweight, I get disc corruption message while gaming after 10-20 minutes randomly. And I can bet that I did not even use 6GB of RAM with all apps running but it still happened. And I used to have the phone rebooting randomly while enumerating games list too. But these issues are gone with the mentioned fixes at least for me.

**Comment by chenxiaolong** [comment]
I just worked around the "reboot" problem in one of my apps. The watchdog in `system_server` uses `IVold.monitor()` to check the state of vold, so if it doesn't respond quickly, the process will be killed and restarted. The problem is that `IVold.monitor()` holds the main vold lock, so any other operation that also holds the main lock will cause the function to block.

Unfortunately, with how Android works right now, a 3rd party app can cause the main vold lock to be held in (at least) two ways:

* When all copies of the proxy file descriptor are closed, FUSE will invoke the `onRelease()` callback. `IVold.unmountAppFuse()`, which holds the main vold lock, is blocked until that returns. I worked around that in my app by performing any on-file-close work in a background thread. (It's impossible to make `close()` blocking in FUSE anyway, so performing synchronous work in `onRelease()` isn't all that useful.)
* When creating a new proxy file descriptor, this makes a call to `IVold.openAppFuseFile()`, which also holds the main vold lock. As part of that call, a `FUSE_LOOKUP` request is performed, which involves calling the `onGetSize()` callback. If the app only uses a single `HandlerThread` for all I/O, then `onGetSize()` can be blocked by an existing slow read/write/flush/getSize for a different file, which ultimately also causes `IVold.monitor()` to block for too long and crash `system_server`. This one can be worked around by just using a new `HandlerThread` for every open file and stopping the thread in `onRelease()`.

**Comment by RebelliousX** [comment]
@chenxiaolong just enable notifications from in app settings and problem is solved, no more reboots.

**Comment by chenxiaolong** [comment]
It's certainly necessary to do that, but it doesn't address the underlying problem. It just avoids the most common reason for it to happen: FUSE callbacks being blocked due to a stopped process. There are still other ways for the callbacks to cause the main vold lock to be held long enough to trigger the watchdog kill.

**Comment by RebelliousX** [comment]
@chenxiaolong I just tried your RSAF latest build, a wonderful app. While I did not encounter reboot issue with RSAF, I encountered file read errors (like in PPSSPP emulator I get disc corrupted message). I highly doubt it is a real issue regarding reading the file, from my experience, this means the app was killed in the background after prolonged gameplay. 

This happens even when battery optimization is disabled (set to unrestricted for RSAF). And somehow, I just tried RSAF, it consumed substantial battery percentage within an hour, around 15 percent while gaming (at least that what Android is saying).

CIFS Document Provider was able to solve this issue by using WorkManger that creates a background service (that creates a foreground service under the hood) that prevents the app from crashing when running in background. The WorkManager only starts when there is at least 1 file opened, and when number files opened is 0, the work ends, thus no background service.  

We also decided on a timeout of 5 seconds when number of files opened reaches 0 to decide for the work to end. This is similar to Debouncing (Debouncing which you may know is a technique used for UI refreshes, which enforces that a function will not be called again until a certain amount of time has passed without it being called., like scrolling too quickly and wait for refreshing until user stops scrolling). By doing this, if opening and closing too many files consecutively and very quickly will not create too many I/O threads, services..etc and closing them, starving thread pool.

And the cool thing about this workaround that **wa2c** was able to come up with instead of my naive foreground service approach, is that we don't even need to disable battery optimization for this to work. The use of WorkManager makes the app very resilient to be terminated by Android OS.  Battery usage of CIFS Document Provider is very low, less than 3 percent even after prolonged emulators gaming sessions.

If you are interested, I would love to see this implemented in your app since it covers more connections scenarios using R-Clone than just CIFS Docuemtn Provider which only offers SMB/SFTP/FTP/FTPS.  Please take a look at this old branch https://github.com/wa2c/cifs-documents-provider/commits/feature/provider_service/

And combined all changes in PR #54 

I don't want to hijack this issue in this repo further. Should I create an issue in your app's repo for this?
